start_admin.sh

```diff
@@ -212,7 +212,34 @@ install_proxy_profile
 
 enable_systemd_services
 
+systemd_options=()
+
+# cgroup v2 compatibility crimes: systemd 219 in the admin container only
+# supports cgroup v1.
+if grep ' /sys/fs/cgroup cgroup2 ' /proc/self/mounts; then
+    # Mount an extra cgroup v1 hierarchy for use by systemd in the admin
+    # container. Vanilla systemd is hard-wired to look for it at /sys/fs/cgroup
+    # but that path is already taken by the host's proper cgroup hierarchy.
+    # Mounting a cgroup v1 hierarchy here might confuse the host and the systemd
+    # documentation advises against manually interfering with anything in that
+    # path. Therefore, mount the extra hierarchy elsewhere and tell systemd in
+    # the admin container where to look. Requirements:
+    #
+    #   1. The base must be a mount point.
+    #   2. The base must contain another mounted cgroup file system named "systemd".
+    #
+    # In either case there is no need to populate the file systems with any
+    # actual cgroup controllers, since the host is expected to manage processes.
+    readonly cgroup_base=/bottlerocket-admin/cgroup
+    mkdir -p "${cgroup_base}"
+    mount -t tmpfs -o nosuid,nodev,noexec tmpfs "${cgroup_base}"
+    mkdir -p "${cgroup_base}"/systemd
+    mount -o remount,ro "${cgroup_base}"
+    mount -t cgroup -o none,name=systemd cgroup "${cgroup_base}"/systemd
```

Comment:
When this mount is added, all processes on the system will have it appear in `/proc/<PID>/cgroup`. Does the entry go away if the admin container is subsequently disabled and this is unmounted?

Reply:
Now this is a bit peculiar. It actually does not, and does not match my expectations (e.g. compare with performing the equivalent steps via `sudo unshare -m`). I still need to dig into this some more. It seems to relate to how the admin container is disabled (e.g. if systemd exits voluntarily before entering its main loop, the mount is cleaned up). So far I could not find any process remaining in the admin container mount namespace and don't understand while the cgroup root seems to remain.

Reply:
I understand this better now. When systemd goes to remove the cgroup of the `sshd.service`, its call to `unlinkat` returns `EBUSY`. The cgroup is never removed. The cgroup root lingers on as long as there's still a cgroup in it, even if it's empty. Even if there doesn't seem to be a way to get back into the mount namespace. Aside from figuring out why removing the cgroup fails, this could also be cleaned up later by remounting the cgroup root with the same name--it'll get reused--and reaping empty cgroups.

While looking at this I found something I had completely overlooked so far, however, and I'll likely have to scrap the entire implementation. :disappointed: See my comment in the main discussion below.

Reply:
An update on the lingering cgroup root. There are two issues at play here:

1. The empty cgroup for `sshd.service` not being removed on exit. The reason for this turns out to be #62. When systemd fails to start `exit.target`, it makes a dash for the exit. Fixing this issue leads to a proper exit and all cgroups being removed.
1. A much more interesting timing-dependent kernel behavior. I'm dedicating the rest of the comment to this one.

Unmounting a cgroup file system can trigger removal of a cgroup root if it was the last mount of that root and the root is otherwise unused. Since with 1. out of the way all cgroups are removed before the container exits, leaving its mount namespace should trigger removal of the cgroup root. However, since the removal of an individual cgroup is an asynchronous multi-step endeavor, unmount and cgroup removal can actually race, with the unmount happening before a cgroup has been fully removed. Once the `rmdir` syscall returns, user-visible effects of cgroup removal have occured, but internal bookkeeping tasks are not yet complete and happen asynchronously. In this case, the cgroup root lingers on and can be observed in e.g. `/proc/1/cgroup` even after the admin container exited.

The issue can be trivially reproduced with e.g. the following sequence of commands (note that `rmdir` and `umount` will need to happen in short succession, so it's best to run this as a script):

```
mkdir -p /mnt/cgroot
mount -t cgroup -o none,name=cgroot cgroup /mnt/cgroot
mkdir /mnt/cgroot/cg
rmdir /mnt/cgroot/cg
umount /mnt/cgroot
```

I spent some time looking at fixing this from the kernel side, but the highly asynchronous nature of the cgroup release code makes this a delicate matter with plenty opportunity to get things wrong. Knowing what to look for now, I noticed I haven't been the first one to observe or try to fix this one either. [The discussion on the cgroups mailing list](https://lore.kernel.org/lkml/20220404142535.145975-1-minhquangbui99@gmail.com/) concluded not to fix this race.

For Bottlerocket's purposes, this is a cosmetic issue. Nothing is really being leaked. The same cgroup root is available and gets reused when the admin container is started again. I'd therefore propose to just live with that for the time being (a workaround would be to re-mount/unmount the hierarchy after some time, e.g. in a `host-ctr` clean-up hook).